Make test file patterns configurable via env vars

jorgemanrubia opened the pull request:
This makes test file patterns configurable via two environment variables: `DEFAULT_TEST`, to configure files to test, and `DEFAULT_TEST_EXCLUDE`, to configure files to exclude from testing.

These values were hardcoded before, which made it difficult to add new categories of tests that should not be executed by default (e.g: smoke tests).

It uses environment variables instead of regular Rails config options because Rails environment is not available when the Runner builds the list of files to test (unless using Spring). A nicer solution would be making sure that the Rails environment is always loaded when the runner starts. This is a first simple step to make these paths configurable for now.

kaspth commented on railties/lib/rails/test_unit/runner.rb:
```diff
@@ -45,7 +45,7 @@ def load_tests(argv)
 
           tests = Rake::FileList[patterns.any? ? patterns : "test/**/*_test.rb"]
-          tests.exclude("test/system/**/*", "test/dummy/**/*") if patterns.empty?
+          tests.exclude(ActiveSupport.excluded_test_paths) if patterns.empty?
```
Can't remember if there was a discussion back when we added this, but curious what it would look like if we flipped it? Such that we only recognize `test/models`, `test/controllers` by default? Or perhaps only run tests for corresponding directories in the app by default? That would set us up with the argument that `rails test` only runs `app/` tests by default (and others' can be run with `rails test test/system` etc.).

Also, doesn't seem right that we just reference `ActiveSupport` here.

jorgemanrubia replied:
> Or perhaps only run tests for corresponding directories in the app by default? That would set us up with the argument that rails test only runs app/ tests by default (and others' can be run with rails test test/system etc.).

I like this idea. But couldn't that be a breaking change for a lot existing code (for people not following those conventions with test folders)?

jorgemanrubia replied:
> Also, doesn't seem right that we just reference ActiveSupport here.

@kaspth maybe better moving the constant to `Rails::TestUnit::Runner` and initialize it in `TestUnitRailtie`?

jorgemanrubia commented:
So the original take was only working when Spring was enabled. 

The problem is that, unless using Spring or other pre-loader, Rails hasn't loaded when the test runner builds the list of files to test. This is done when requiring the "test_helper.rb" in the first test. This means, that using regular rails config options and initialization callbacks won't work for tweaking the list of test paths in the general case.

I've found 2 alternatives you can see in 2 separated commits:

1. [Simply extract the list of paths to a property in `Rails::TestUnit::Runner`](https://github.com/rails/rails/pull/39036/commits/5c21ea4cd34ce097b27172bdd33599e116594340
). If you are using Spring, you can still configure this value in an initializer or in `config/environment/test.rb`. 

2. [Make the test command require the environment and rails app before executing](https://github.com/rails/rails/pull/39036/commits/a17e4a1039a8f65c3531131d680aa0fa24aed6c9). We would need to do the same for supporting `rake` task execution. 

I'd go with (1), it's way simpler and doesn't introduce the big change of making test commands require the Rails environment before executing.

jeremy commented:
1. A Spring-only feature is confusing for general usage. Don't think that'll fly.
2. That's surprising! I am curious to what degree this is intentional.

Another alternative could be to allow apps to override the default test globs with environment variables. Low-level, but start by making it possible!

```diff
diff --git a/railties/lib/rails/test_unit/runner.rb b/railties/lib/rails/test_unit/runner.rb
index 0e7ce5afb3..4538090615 100644
--- a/railties/lib/rails/test_unit/runner.rb
+++ b/railties/lib/rails/test_unit/runner.rb
@@ -8,6 +8,9 @@
 module Rails
   module TestUnit
     class Runner
+      DEFAULT_TEST_GLOB = "test/**/*_test.rb"
+      DEFAULT_TEST_EXCLUDE_GLOB = "test/{system,dummy}/**/*_test.rb"
+
       mattr_reader :filters, default: []
 
       class << self
@@ -42,12 +45,7 @@ def run(argv = [])
         end
 
         def load_tests(argv)
-          patterns = extract_filters(argv)
-
-          tests = Rake::FileList[patterns.any? ? patterns : "test/**/*_test.rb"]
-          tests.exclude("test/system/**/*", "test/dummy/**/*") if patterns.empty?
-
-          tests.to_a.each { |path| require File.expand_path(path) }
+          test_paths(argv).to_a.each { |path| require File.expand_path(path) }
         end
 
         def compose_filter(runnable, filter)
@@ -59,6 +57,28 @@ def compose_filter(runnable, filter)
         end
 
         private
+          def test_paths(argv)
+            filtered_test_paths(argv) || default_test_paths
+          end
+
+          def default_test_paths
+            Rake::FileList[default_test_glob].exclude(default_test_exclude_glob)
+          end
+
+          def default_test_glob
+            ENV.fetch("DEFAULT_TEST", DEFAULT_TEST_GLOB)
+          end
+
+          def default_test_exclude_glob
+            ENV.fetch("DEFAULT_TEST_EXCLUDE", DEFAULT_TEST_EXCLUDE_GLOB)
+          end
+
+          def filtered_test_paths(argv)
+            if patterns = extract_filters(argv).presence
+              Rake::FileList[patterns]
+            end
+          end
+
           def extract_filters(argv)
             # Extract absolute and relative paths but skip -n /.*/ regexp filters.
             argv.select { |arg| path_argument?(arg) && !regexp_filter?(arg) }.map do |path|
```

jorgemanrubia committed:
Make test file patterns configurable via Environment variables

This makes test file patterns configurable via two environment variables:
`DEFAULT_TEST`, to configure files to test, and `DEFAULT_TEST_EXCLUDE`,
to configure files to exclude from testing.

These values were hardcoded before, which made it difficult to add
new categories of tests that should not be executed by default (e.g:
smoke tests).

It uses environment variables instead of regular Rails config options
because Rails environment is not available when the Runner builds the
list of files to test (unless using Spring). A nicer solution would be
making sure that the Rails environment is always loaded when the runner
starts. This is a first simple step to make these paths configurable for
now

This way at least you could override defaults in `config/boot.rb`:

```ruby
ENV["DEFAULT_TEST_EXCLUDE"] = "test/{dummy,smoke,system}/**/*_test.rb
```

Co-authored-by: Jeremy Daer <jeremydaer@gmail.com>

jorgemanrubia commented:
Thanks @Jeremy. Implemented that approach and added some tests 👍.

jorgemanrubia commented on railties/lib/rails/test_unit/runner.rb:
```diff
@@ -59,6 +57,27 @@ def compose_filter(runnable, filter)
+
+          def filtered_test_paths(argv)
+            patterns = extract_filters(argv)
+            Rake::FileList[patterns] if patterns.any?
```
`#presence` not available since you can't count on Rails being loaded

kaspth replied:
We actually need to add a missing `require "active_support"` up top (we require callers to load the top level file to spare internal require calls). Then `presence` is available I think.

kaspth commented on railties/lib/rails/test_unit/runner.rb:
```diff
@@ -59,6 +57,27 @@ def compose_filter(runnable, filter)
+          end
+
+          def default_test_exclude_glob
+            ENV.fetch("DEFAULT_TEST_EXCLUDE", DEFAULT_TEST_EXCLUDE_GLOB)
```
This is a public class so the constants will be added to the public API. Don't think we want that. Since we've already extracted to a method, why don't we inline them and use standard `[] ||` as well?

kaspth commented on railties/lib/rails/test_unit/runner.rb:
```diff
@@ -42,12 +45,7 @@ def run(argv = [])
-          tests.exclude("test/system/**/*", "test/dummy/**/*") if patterns.empty?
-
-          tests.to_a.each { |path| require File.expand_path(path) }
+          test_paths(argv).to_a.each { |path| require File.expand_path(path) }
```
I don't think the extra layers of indirection buys us much here, we're laying more concrete that it's suggested we want to mold further in another release.

Let's start even simpler and just extract the 2 patterns ala:

```ruby
def load_tests(argv)
  patterns = extract_filters(argv)

  tests = Rake::FileList[patterns.presence || default_test_glob]
  tests.exclude(default_test_exclude_glob) if patterns.empty?

  tests.to_a.each { |path| require File.expand_path(path) }
end

def default_test_glob
  ENV["DEFAULT_TEST"] || "test/**/*_test.rb"
end

def default_test_exclude_glob
  ENV["DEFAULT_TEST_EXCLUDE"] || "test/{system,dummy}/**/*_test.rb"
end
```

jorgemanrubia commented:
Thanks @kaspth!

I like that version that @jeremy cooked with extracted methods and constants, but no problem in making those changes. I'll do tomorrow

kaspth commented on railties/lib/rails/test_unit/runner.rb:
```diff
@@ -44,12 +45,20 @@ def run(argv = [])
           tests.to_a.each { |path| require File.expand_path(path) }
         end
 
+        def default_test_glob
+          ENV["DEFAULT_TEST"] || "test/**/*_test.rb"
+        end
+
+        def default_test_exclude_glob
+          ENV["DEFAULT_TEST_EXCLUDE"] || "test/{system,dummy}/**/*_test.rb"
+        end
```
Still think these should be private.

rafaelfranca commented:
I'm not sure I get the reason for this PR.

What is the reason to do

`DEFAULT_TEST_EXCLUDE="test/smoke/**/*_test.rb" bin/rails test`

and not

`bin/rails test test/smoke/**/*_test.rb`?

rafaelfranca commented on railties/lib/rails/test_unit/runner.rb:
```diff
@@ -44,12 +45,20 @@ def run(argv = [])
 
-          tests = Rake::FileList[patterns.any? ? patterns : "test/**/*_test.rb"]
-          tests.exclude("test/system/**/*", "test/dummy/**/*") if patterns.empty?
+          tests = Rake::FileList[patterns.presence || default_test_glob]
```
Do we really need to use `presence` here? The previous code were just fine.

jorgemanrubia replied:
@jeremy added it in his initial suggestion, I removed it since no active support extension, and then @kaspth suggested adding active support extensions so I did 😂. I'll remove it

kaspth commented on railties/test/application/test_runner_test.rb:
```diff
@@ -895,6 +895,20 @@ class DummyTest < ApplicationSystemTestCase
+    def test_can_exclude_files_from_being_tested_via_default_rails_command_by_setting_DEFAULT_TEST_EXCLUDE_env_var
+      create_test_file "smoke", "smoke_foo"
+
+      output = Dir.chdir(app_path) { `DEFAULT_TEST_EXCLUDE="test/smoke/**/*_test.rb" bin/rails test` }
```
Can we use `run_test_command` wrapped in `switch_env`?

Same as what we do here: https://github.com/rails/rails/blob/f4538aa5865da5f16b707d34aa733a5e2353f2fd/railties/test/commands/secrets_test.rb#L64-L68

jorgemanrubia replied:
I looked for something like `switch_env` but couldn't find it, thanks 👍

What I liked about the previous approach is that both tests (for `rake test` and `rails t`) were symmetrical. But this is more consistent with other similar tests.

kaspth commented:
@rafaelfranca `bin/rails test` would automatically include `test/smoke` because only system and dummy tests are excluded by default.

I still think there's a way to do https://github.com/rails/rails/pull/39036#discussion_r414652705, but it's complicated because the Rails application hasn't been loaded then so a new_framework_default isn't accessible, say.

jorgemanrubia commented:
Yes @rafaelfranca, our use case is wanting to exclude a folder in addition to `/system`, so that it's excluded when running tests via `rails test` or `rake test`. Right now the list of paths is hard-coded. 

Using env vars instead of regular rails config because the Rails environment isn't loaded when the runner is calculating the list of paths, unless Spring is used

jorgemanrubia commented:
Thanks for your suggestions @kaspth and @rafaelfranca. I think I addressed all the comments. This is good to merge from my side.

kaspth commented:
Thanks @jorgemanrubia! I merged in d143154147 because after the switch to `switch_env` both tests referred to `bin/rails`.
